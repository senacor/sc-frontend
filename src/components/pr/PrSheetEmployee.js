import React from 'react';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles/index';
import { debounce } from '../../helper/debounce';
import { connect } from 'react-redux';
import * as actions from '../../actions';

const styles = theme => ({
  bootstrapInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '10px 12px',
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  nestedText: {
    paddingLeft: '30px',
    paddingRight: '30px'
  },
  nestedListItem: {
    paddingLeft: '30px',
    width: '93%'
  }
});

class PrSheetEmployee extends React.Component {
  constructor(props) {
    super(props);

    let reflectionSet = this.props.prById.prReflectionSet.find(
      prReflection => prReflection.prReflectionField === this.props.category
    );

    this.state = {
      text: reflectionSet ? reflectionSet.text : ''
    };
  }

  handleChangeComment = (prById, category) => event => {
    this.setState({ text: event.target.value });

    let reflectionSet = prById.prReflectionSet.find(
      prReflection => prReflection.prReflectionField === category
    );

    this.sendComment(prById, category, event.target.value, reflectionSet.id);
  };

  sendComment = debounce(this.props.addEmployeeContribution, 500);

  translateCategoryName = categoryName => {
    switch (categoryName) {
      case 'INFLUENCE_OF_LEADER_AND_ENVIRONMENT':
        return 'Reflektion Beitrag des Vorgesetzten und Umfeld zur eigenen Entwicklung';
      case 'ROLE_AND_PROJECT_ENVIRONMENT':
        return 'Rolle und Projektumfeld';
      default:
        return 'Reflektion Beitrag des Vorgesetzten und Umfeld zur eigenen Entwicklung';
    }
  };

  placeholder = categoryName => {
    switch (categoryName) {
      case 'ROLE_AND_PROJECT_ENVIRONMENT':
        return 'Beschreibung Projekt, Aufgabe und Rolle';
      case 'INFLUENCE_OF_LEADER_AND_ENVIRONMENT':
        return 'Was ist mir f체r meine Entwicklung wichtig? Welchen Beitrag leistet mein Umfeld und mein Vorgesetzter, was sollte sich 채ndern und wie kann ich dazu beitragen?';
      default:
        return 'Was ist mir f체r meine Entwicklung wichtig? Welchen Beitrag leistet mein Umfeld und mein Vorgesetzter, was sollte sich 채ndern und wie kann ich dazu beitragen?';
    }
  };

  render() {
    const { prById, category, classes } = this.props;
    return (
      <div>
        <ListItem className={classes.nestedListItem}>
          <ListItemText secondary={this.translateCategoryName(category)} />
        </ListItem>
        <List component="div" disablePadding className={classes.nestedText}>
          <ListItem>
            <TextField
              id={category + '_CommentId'}
              multiline
              fullWidth
              rowsMax="4"
              value={this.state.text ? this.state.text : ''}
              onChange={this.handleChangeComment(prById, category)}
              InputProps={{
                disableUnderline: true,
                name: 'comment',
                classes: {
                  input: classes.bootstrapInput
                }
              }}
              InputLabelProps={{
                shrink: true
              }}
            />
          </ListItem>
          <ListItem>
            <ListItemText secondary={this.placeholder(category)} />
          </ListItem>
        </List>
      </div>
    );
  }
}

export const StyledComponent = withStyles(styles)(PrSheetEmployee);
export default connect(
  state => ({
    prEmployeeContribution: state.prEmployeeContributions.prEmployeeContribution
  }),
  {
    addEmployeeContribution: actions.addEmployeeContribution
  }
)(StyledComponent);
