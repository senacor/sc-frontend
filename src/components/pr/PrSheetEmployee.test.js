import React from 'react';
import { StyledComponent } from './PrSheetEmployee';
import { createShallow } from '@material-ui/core/test-utils';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';

describe('PrSheetEmployee Component', () => {
  let shallow = createShallow({ dive: true });

  const prById = {
    id: 1,
    supervisor: 'ttran',
    occasion: 'ON_DEMAND',
    status: 'PREPARATION',
    deadline: '2018-03-14',

    prReflectionSet: [
      {
        id: 1,
        prReflectionField: 'ROLE_AND_PROJECT_ENVIRONMENT',
        text: null
      },
      {
        id: 2,
        prReflectionField: 'INFLUENCE_OF_LEADER_AND_ENVIRONMENT',
        text: 'hjkkkhgcghjjn'
      }
    ],

    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/prs/1'
      }
    }
  };

  it('displays the PrSheetEmployee, component: ROLE_AND_PROJECT_ENVIRONMENT ', () => {
    const component = shallow(
      <StyledComponent
        prById={prById}
        category="ROLE_AND_PROJECT_ENVIRONMENT"
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('displays the PrSheetEmployee, component: INFLUENCE_OF_LEADER_AND_ENVIRONMENT', () => {
    const component = shallow(
      <StyledComponent
        prById={prById}
        category="INFLUENCE_OF_LEADER_AND_ENVIRONMENT"
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('displays the text read-only for the supervisor', () => {
    const component = shallow(
      <StyledComponent
        prById={prById}
        category="ROLE_AND_PROJECT_ENVIRONMENT"
        userroles={['PR_CST_Leiter']}
      />
    );

    expect(component.find(Typography)).toHaveLength(1);
    expect(component.find(TextField)).toHaveLength(0);
  });
});
