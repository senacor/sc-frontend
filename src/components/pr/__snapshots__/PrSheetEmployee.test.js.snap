// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrSheetEmployee Component displays the PrSheetEmployee, component: INFLUENCE_OF_LEADER_AND_ENVIRONMENT 1`] = `
<div>
  <WithStyles(ListItem)
    className="PrSheetEmployee-nestedListItem-3"
  >
    <WithStyles(ListItemText)
      secondary="Reflektion Beitrag des Vorgesetzten und Umfeld zur eigenen Entwicklung"
    />
  </WithStyles(ListItem)>
  <WithStyles(List)
    className="PrSheetEmployee-nestedText-2"
    component="div"
    disablePadding={true}
  >
    <WithStyles(ListItem)>
      <TextField
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        InputProps={
          Object {
            "classes": Object {
              "input": "PrSheetEmployee-bootstrapInput-1",
            },
            "disableUnderline": true,
            "name": "comment",
          }
        }
        fullWidth={true}
        id="INFLUENCE_OF_LEADER_AND_ENVIRONMENT_CommentId"
        multiline={true}
        onChange={[Function]}
        required={false}
        rowsMax="4"
        select={false}
        value="hjkkkhgcghjjn"
      />
    </WithStyles(ListItem)>
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        secondary="Was ist mir für meine Entwicklung wichtig? Welchen Beitrag leistet mein Umfeld und mein Vorgesetzter, was sollte sich ändern und wie kann ich dazu beitragen?"
      />
    </WithStyles(ListItem)>
  </WithStyles(List)>
</div>
`;

exports[`PrSheetEmployee Component displays the PrSheetEmployee, component: ROLE_AND_PROJECT_ENVIRONMENT  1`] = `
<div>
  <WithStyles(ListItem)
    className="PrSheetEmployee-nestedListItem-3"
  >
    <WithStyles(ListItemText)
      secondary="Rolle und Projektumfeld"
    />
  </WithStyles(ListItem)>
  <WithStyles(List)
    className="PrSheetEmployee-nestedText-2"
    component="div"
    disablePadding={true}
  >
    <WithStyles(ListItem)>
      <TextField
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        InputProps={
          Object {
            "classes": Object {
              "input": "PrSheetEmployee-bootstrapInput-1",
            },
            "disableUnderline": true,
            "name": "comment",
          }
        }
        fullWidth={true}
        id="ROLE_AND_PROJECT_ENVIRONMENT_CommentId"
        multiline={true}
        onChange={[Function]}
        required={false}
        rowsMax="4"
        select={false}
        value=""
      />
    </WithStyles(ListItem)>
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        secondary="Beschreibung Projekt, Aufgabe und Rolle"
      />
    </WithStyles(ListItem)>
  </WithStyles(List)>
</div>
`;
